### ISSUE TITLE
Mergeable hangs on my config

### ISSUE THREAD

**Issue by pkuczynski** [mergeability/mergeable#330]
Used to be super fast, but now runs for >5min and no response. Is it my config or something on your side?

This is my repo: https://github.com/rvm/rvm/pull/4963/checks?check_run_id=858044676

**Comment by shine2lay** [comment]
@pkuczynski I believe i see the issue here, It's an weird edge case we haven't addressed yet. We will address it in the near future.
I think the issue is with the fail case:
![image](https://user-images.githubusercontent.com/5243508/87174831-a359c200-c28c-11ea-8347-ccaca409d6cf.png)

You'll need to put the `check` action in fail case. 
By default, we use `check` action if `pass,fail,error` are undefined. if it's defined, it overwrite the defaults.

**Comment by pkuczynski** [comment]
And how would that look like? Can you give me an example, please?

**Comment by shine2lay** [comment]
yes, here you go:

```
fail:
  - do : assign
    assignees: ['@author']
  - do: checks # default fail case
  status: 'failure' # Can be: success, failure, neutral, cancelled, timed_out, or action_required
  payload:
    title: 'Mergeable Run have been Completed!'
    summary: |
         ### Status: {{toUpperCase validationStatus}}
              Here are some stats of the run:
              {{validationCount}} validations were ran.
              {{passCount}} PASSED
              {{failCount}} FAILED
    text: "{{#each validationSuites}}\n
          #### {{{statusIcon status}}} Validator: {{toUpperCase name}}\n
          {{#each validations }} * {{{statusIcon status}}} ***{{{ description }}}***\n
                 Input : {{{details.input}}}\n
                 Settings : {{{displaySettings details.settings}}}\n
                 {{/each}}\n
          {{/each}}"
```

**Comment by pkuczynski** [comment]
Do I need to use the payload? Would this be enough?

```
    fail:
      - do: assign
        assignees: [ '@author' ]
      - do: checks
        status: 'failure'
```

**Comment by pkuczynski** [comment]
All I want to achieve is, that in case of the missing assignment it would assign the PR author. Maybe there is a better way of doing this?

**Comment by shine2lay** [comment]
> All I want to achieve is, that in case of the missing assignment it would assign the PR author. Maybe there is a better way of doing this?

How about splitting this use case into it's own recipe, something like this: 
```
- when: pull_request.*
  name: "check missing assignment"
  validate:
    - do : assignee
      no_empty:
        enabled: true
  pass:
  fail:
    - do: assign
      assignees: [ '@author' ]
  error:
```

**Comment by pkuczynski** [comment]
Just did that: https://github.com/rvm/rvm/pull/4963/files#diff-062bf34748193dedcee2ec2a47fed4e5R5

Do I always have to put `pass` and `error` to silently make the change? What if the author is outside of the org?

**Comment by shine2lay** [comment]
> Do I always have to put `pass` and `error` to silently make the change? 
Unfortunately, if you don't, you may have a case where a check is hanged. Until we addressed the edge case i mentioned above.

> What if the author is outside of the org?
good question, I don't know the answer to this. I'll try it out and let you know.

**Comment by pkuczynski** [comment]
Unfortunately, your suggestion with empty `pass` and `error` fails mergable: https://github.com/rvm/rvm/pull/4963/checks?check_run_id=859406735

**Comment by pkuczynski** [comment]
> Edit: it seems like it won't work if the author is outside of the org , [rvm/rvm#4964](https://github.com/rvm/rvm/pull/4964) .

Would it fail loudly or silently?

**Comment by shine2lay** [comment]
> Unfortunately, your suggestion with empty `pass` and `error` fails mergable: https://github.com/rvm/rvm/pull/4963/checks?check_run_id=859406735

hmmm.. i haven't seen this error before, let me try to debug it

> Would it fail loudly or silently?
It seemed to fail silently, we have a general error handler that should comment if an unexpected error occurred. Would it be helpful if it fails loudly?

**Comment by pkuczynski** [comment]
I would prefer it fails silently :) If it can't assign then, don't assign...

**Comment by shine2lay** [comment]
> Unfortunately, your suggestion with empty `pass` and `error` fails mergable: https://github.com/rvm/rvm/pull/4963/checks?check_run_id=859406735

It seems that some of my new code breaks this case so this is no longer a solution to the edge case.
if you don't mind waiting a few days, i'll try to just address this edge case and your other requirements.

**Comment by shine2lay** [comment]
if you have other requirements, do let me know so i can try to accommodate it as best as i can.

**Comment by pkuczynski** [comment]
@shine2lay now mergable is completely gone from my repo!?! 

https://github.com/rvm/rvm/pull/4963

**Comment by shine2lay** [comment]
@pkuczynski hi, good news and bad news.
Good news is that the fix is working as expected. implemented default flow can be found #332 
Bad news is that, it surfaced a differently unrelated bug that we weren't aware of. #337 

Tldr; we can't get rid of checks in multiple recipe format yet.

Sorry about the in convinces. I'll work on fixing the new bug
Below are two config (I've tested both of them) that should work.
```
defaults:
  checks_failure_payload: &checks_failure_payload
    title: 'Mergeable Run have been Completed!'
    summary: |
      ### Status: {{toUpperCase validationStatus}}
          Here are some stats of the run:
          {{validationCount}} validations were ran.
          {{passCount}} PASSED
          {{failCount}} FAILED
    text: "{{#each validationSuites}}\n
          #### {{{statusIcon status}}} Validator: {{toUpperCase name}}\n
          {{#each validations }} * {{{statusIcon status}}} ***{{{ description }}}***\n
                  Input : {{{details.input}}}\n
                  Settings : {{{displaySettings details.settings}}}\n
                  {{/each}}\n
          {{/each}}"

version: 2
mergeable:
  # Validate assignee
  - when: pull_request.*
    name: "assignee"
    validate:
      - do: assignee
        min:
          count: 1
    fail:
      - do: checks
        status: 'failure'
        payload: *checks_failure_payload
      - do: assign
        assignees: [ '@author' ]

  # Validate fields changeable by maintainers
  - when: pull_request.*
    name: "triage"
    validate:
      - do: label
        and:
          - no_empty:
              enabled: true
          - must_exclude:
              regex: duplicate|feedback|invalid|wont|progress
      - do: milestone
        no_empty:
          enabled: true

  # Validate fields changeable by anybody
  - when: pull_request.*
    name: "content"
    validate:
      - do: description
        no_empty:
          enabled: true
          message: Description should not be empty. Provide details with **what** was changed, **why** it was changed, and **how** it was changed.
      - do: dependent
        changed:
          file: '*'
          files: ['CHANGELOG.md']
        message: Missing CHANGELOG entry
      - do: dependent
        changed:
          file: 'binscripts/rvm-installer'
          files: ['binscripts/rvm-installer.asc']
        message: Installer is missing updated signature
      - do: dependent
        changed:
          file: 'config/known'
          files: ['config/known_strings', 'config/md5', 'config/sha512']
        message: New ruby version needs to be listed as known and with signatures
      - do: dependent
        changed:
          file: 'config/known_strings'
          files: ['config/md5', 'config/sha512']
        message: New ruby version needs to be listed as known and with signatures
      - do: dependent
        files: ['config/md5', 'config/sha512']
        message: Both md5 and sha512 signatures needs to be modified together
    fail:
      - do: checks
        status: 'failure'
        payload: *checks_failure_payload
      - do: comment
        payload:
          body: This PR does not contain all required information and changes. Check the `mergeable` status for more details.

  # Stale PR
  - when: schedule.repository
    validate:
      - do: stale
        days: 20
        type: pull_request, issues
    pass:
      - do: comment
        payload:
          body: There was no activity for 20 days. Is this still relevant?

```

```
defaults:
  checks_failure_payload: &checks_failure_payload
    title: 'Mergeable Run have been Completed!'
    summary: |
      ### Status: {{toUpperCase validationStatus}}
          Here are some stats of the run:
          {{validationCount}} validations were ran.
          {{passCount}} PASSED
          {{failCount}} FAILED
    text: "{{#each validationSuites}}\n
          #### {{{statusIcon status}}} Validator: {{toUpperCase name}}\n
          {{#each validations }} * {{{statusIcon status}}} ***{{{ description }}}***\n
                  Input : {{{details.input}}}\n
                  Settings : {{{displaySettings details.settings}}}\n
                  {{/each}}\n
          {{/each}}"

version: 2
mergeable:
  - when: pull_request.*
    name: "triage"
    validate:
      - do: label
        and:
          - no_empty:
              enabled: true
          - must_exclude:
              regex: duplicate|feedback|invalid|wont|progress
      - do: milestone
        no_empty:
          enabled: true
      - do: assignee
        min:
          count: 1
      - do: description
        no_empty:
          enabled: true
          message: Description should not be empty. Provide details with **what** was changed, **why** it was changed, and **how** it was changed.
      - do: dependent
        changed:
          file: '*'
          files: ['CHANGELOG.md']
        message: Missing CHANGELOG entry
      - do: dependent
        changed:
          file: 'binscripts/rvm-installer'
          files: ['binscripts/rvm-installer.asc']
        message: Installer is missing updated signature
      - do: dependent
        changed:
          file: 'config/known'
          files: ['config/known_strings', 'config/md5', 'config/sha512']
        message: New ruby version needs to be listed as known and with signatures
      - do: dependent
        changed:
          file: 'config/known_strings'
          files: ['config/md5', 'config/sha512']
        message: New ruby version needs to be listed as known and with signatures
      - do: dependent
        files: ['config/md5', 'config/sha512']
        message: Both md5 and sha512 signatures needs to be modified together
    fail:
      - do: checks
        status: 'failure'
        payload: *checks_failure_payload
      - do: assign
        assignees: [ '@author' ]
```

**Comment by pkuczynski** [comment]
@shine2lay sorry for looking into this so late. Do I still need to make the changes with the payload?

**Comment by shine2lay** [comment]
@pkuczynski the one you currently have in https://github.com/rvm/rvm/pull/4963/files will only assign, it will not create checks. if you want to add checks use the following

```
defaults:
  checks_failure_payload: &checks_failure_payload
    title: 'Mergeable Run have been Completed!'
    summary: |
      ### Status: {{toUpperCase validationStatus}}
          Here are some stats of the run:
          {{validationCount}} validations were ran.
          {{passCount}} PASSED
          {{failCount}} FAILED
    text: "{{#each validationSuites}}\n
          #### {{{statusIcon status}}} Validator: {{toUpperCase name}}\n
          {{#each validations }} * {{{statusIcon status}}} ***{{{ description }}}***\n
                  Input : {{{details.input}}}\n
                  Settings : {{{displaySettings details.settings}}}\n
                  {{/each}}\n
          {{/each}}"

version: 2
mergeable:
  - when: pull_request.*
    name: "triage"
    validate:
      - do: label
        and:
          - no_empty:
              enabled: true
          - must_exclude:
              regex: duplicate|feedback|invalid|wont|progress
      - do: milestone
        no_empty:
          enabled: true
      - do: assignee
        min:
          count: 1
      - do: description
        no_empty:
          enabled: true
          message: Description should not be empty. Provide details with **what** was changed, **why** it was changed, and **how** it was changed.
      - do: dependent
        changed:
          file: '*'
          files: ['CHANGELOG.md']
        message: Missing CHANGELOG entry
      - do: dependent
        changed:
          file: 'binscripts/rvm-installer'
          files: ['binscripts/rvm-installer.asc']
        message: Installer is missing updated signature
      - do: dependent
        changed:
          file: 'config/known'
          files: ['config/known_strings', 'config/md5', 'config/sha512']
        message: New ruby version needs to be listed as known and with signatures
      - do: dependent
        changed:
          file: 'config/known_strings'
          files: ['config/md5', 'config/sha512']
        message: New ruby version needs to be listed as known and with signatures
      - do: dependent
        files: ['config/md5', 'config/sha512']
        message: Both md5 and sha512 signatures needs to be modified together
    fail:
      - do: checks
        status: 'failure'
        payload: *checks_failure_payload
      - do: assign
        assignees: [ '@author' ]
```
This one should create `checks` and also attempt to `assign` author when validation fails
